tests/functional/object/mcg/test_custom_credentials_using_mcg_cli.py::TestCustomCredsUsingMcgCli::test_account_update_with_custom_creds_using_cli is failing

tests/functional/object/mcg/test_custom_credentials_using_mcg_cli.py::TestCustomCredsUsingMcgCli::test_account_update_with_custom_creds_using_cli is failing .  
Check on 4.20 and fix if the failure is persistent. 

Relevant log message: 
------------------------------- live log call ---------------------------------
06:28:13 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: oc --kubeconfig /root/auth/kubeconfig -n openshift-storage get csv -n openshift-storage -o yaml
06:28:15 - MainThread - ocs_ci.helpers.odf_cli - INFO  - ODF CLI binary is ready at /root/ocs-upi-kvm/src/ocs-ci/bin/odf
06:28:15 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: /root/ocs-upi-kvm/src/ocs-ci/data/odf-cli noobaa version
06:28:16 - MainThread - ocs_ci.framework.pytest_customization.reports - INFO  - duration reported by tests/functional/object/mcg/test_custom_credentials_using_mcg_cli.py::TestCustomCredsUsingMcgCli::test_account_update_with_custom_creds_using_cli immediately after test execution: 1.96
FAILED
__ TestCustomCredsUsingMcgCli.test_account_update_with_custom_creds_using_cli __

self = <mcg.test_custom_credentials_using_mcg_cli.TestCustomCredsUsingMcgCli object at 0x7ffe95700f40>
mcg_account_factory = <function mcg_account_factory_fixture.<locals>.mcg_account_factory_implementation at 0x7ffe90f621f0>

    @skipif_ocs_version("<4.17")
    @tier2
    def test_account_update_with_custom_creds_using_cli(self, mcg_account_factory):
        """
        1. Create a new account using MCG CLI
        2. Retrive creds of the account
        3. Change creds of the with custom value suing CLI
        4. Verify creds are updated using CLI
        5. Validate custom credentials against against length and valid characters
        """
        if not Path(mcg_cli).exists():
            # Download appropriate CLI binary based on version
            ocs_version = version_module.get_semantic_ocs_version_from_config()
            if ocs_version >= version_module.VERSION_4_20:
                retrieve_cli_binary(cli_type="odf")
            else:
                retrieve_cli_binary(cli_type="mcg")

        # Build version command based on CLI version
        version_cmd = build_mcg_command("version")

>       output = run_cmd(cmd=version_cmd)

tests/functional/object/mcg/test_custom_credentials_using_mcg_cli.py:82:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ocs_ci/utility/utils.py:491: in run_cmd
    completed_process = exec_cmd(
ocs_ci/utility/retry.py:79: in f_retry
    return f(*args, **kwargs)
ocs_ci/utility/utils.py:700: in exec_cmd
    completed_process = subprocess.run(
/usr/lib64/python3.9/subprocess.py:505: in run
    with Popen(*popenargs, **kwargs) as process:
/usr/lib64/python3.9/subprocess.py:951: in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Popen: returncode: 255 args: ['/root/ocs-upi-kvm/src/ocs-ci/data/odf-cli', ...>
args = ['/root/ocs-upi-kvm/src/ocs-ci/data/odf-cli', 'noobaa', 'version']
executable = b'/root/ocs-upi-kvm/src/ocs-ci/data/odf-cli', preexec_fn = None
close_fds = True, pass_fds = (), cwd = None
env = {'BASH_FUNC_which%%': '() {  ( alias;\n eval ${which_declare} ) | /usr/bin/which --tty-only --read-alias --read-functi..._ADDRESS': 'unix:path=/run/user/0/bus', 'DEBUGINFOD_IMA_CERT_PATH': '/etc/keys/ima:', 'HISTCONTROL': 'ignoredups', ...}
startupinfo = None, creationflags = 0, shell = False, p2cread = 8, p2cwrite = 9
c2pread = 10, c2pwrite = 11, errread = 12, errwrite = 13, restore_signals = True
gid = None, gids = None, uid = None, umask = -1, start_new_session = False

    def _execute_child(self, args, executable, preexec_fn, close_fds,
                       pass_fds, cwd, env,
                       startupinfo, creationflags, shell,
                       p2cread, p2cwrite,
                       c2pread, c2pwrite,
                       errread, errwrite,
                       restore_signals,
                       gid, gids, uid, umask,
                       start_new_session):
        """Execute program (POSIX version)"""

        if isinstance(args, (str, bytes)):
            args = [args]
        elif isinstance(args, os.PathLike):
            if shell:
                raise TypeError('path-like args is not allowed when '
                                'shell is true')
            args = [args]
        else:
            args = list(args)

        if shell:
            # On Android the default shell is at '/system/bin/sh'.
            unix_shell = ('/system/bin/sh' if
                      hasattr(sys, 'getandroidapilevel') else '/bin/sh')
            args = [unix_shell, "-c"] + args
            if executable:
                args[0] = executable

        if executable is None:
            executable = args[0]

        sys.audit("subprocess.Popen", executable, args, cwd, env)

        if (_USE_POSIX_SPAWN
                and os.path.dirname(executable)
                and preexec_fn is None
                and not close_fds
                and not pass_fds
                and cwd is None
                and (p2cread == -1 or p2cread > 2)
                and (c2pwrite == -1 or c2pwrite > 2)
                and (errwrite == -1 or errwrite > 2)
                and not start_new_session
                and gid is None
                and gids is None
                and uid is None
                and umask < 0):
            self._posix_spawn(args, executable, env, restore_signals,
                              p2cread, p2cwrite,
                              c2pread, c2pwrite,
                              errread, errwrite)
            return

        orig_executable = executable

        # For transferring possible exec failure from child to parent.
        # Data format: "exception name:hex errno:description"
        # Pickle is not used; it is complex and involves memory allocation.
        errpipe_read, errpipe_write = os.pipe()
        # errpipe_write must not be in the standard io 0, 1, or 2 fd range.
        low_fds_to_close = []
        while errpipe_write < 3:
            low_fds_to_close.append(errpipe_write)
            errpipe_write = os.dup(errpipe_write)
        for low_fd in low_fds_to_close:
            os.close(low_fd)
        try:
            try:
                # We must avoid complex work that could involve
                # malloc or free in the child process to avoid
                # potential deadlocks, thus we do all this here.
                # and pass it to fork_exec()

                if env is not None:
                    env_list = []
                    for k, v in env.items():
                        k = os.fsencode(k)
                        if b'=' in k:
                            raise ValueError("illegal environment variable name")
                        env_list.append(k + b'=' + os.fsencode(v))
                else:
                    env_list = None  # Use execv instead of execve.
                executable = os.fsencode(executable)
                if os.path.dirname(executable):
                    executable_list = (executable,)
                else:
                    # This matches the behavior of os._execvpe().
                    executable_list = tuple(
                        os.path.join(os.fsencode(dir), executable)
                        for dir in os.get_exec_path(env))
                fds_to_keep = set(pass_fds)
                fds_to_keep.add(errpipe_write)
                self.pid = _posixsubprocess.fork_exec(
                        args, executable_list,
                        close_fds, tuple(sorted(map(int, fds_to_keep))),
                        cwd, env_list,
                        p2cread, p2cwrite, c2pread, c2pwrite,
                        errread, errwrite,
                        errpipe_read, errpipe_write,
                        restore_signals, start_new_session,
                        gid, gids, uid, umask,
                        preexec_fn)
                self._child_created = True
            finally:
                # be sure the FD is closed no matter what
                os.close(errpipe_write)

            self._close_pipe_fds(p2cread, p2cwrite,
                                 c2pread, c2pwrite,
                                 errread, errwrite)

            # Wait for exec to fail or succeed; possibly raising an
            # exception (limited in size)
            errpipe_data = bytearray()
            while True:
                part = os.read(errpipe_read, 50000)
                errpipe_data += part
                if not part or len(errpipe_data) > 50000:
                    break
        finally:
            # be sure the FD is closed no matter what
            os.close(errpipe_read)

        if errpipe_data:
            try:
                pid, sts = os.waitpid(self.pid, 0)
                if pid == self.pid:
                    self._handle_exitstatus(sts)
                else:
                    self.returncode = sys.maxsize
            except ChildProcessError:
                pass

            try:
                exception_name, hex_errno, err_msg = (
                        errpipe_data.split(b':', 2))
                # The encoding here should match the encoding
                # written in by the subprocess implementations
                # like _posixsubprocess
                err_msg = err_msg.decode()
            except ValueError:
                exception_name = b'SubprocessError'
                hex_errno = b'0'
                err_msg = 'Bad exception data from child: {!r}'.format(
                              bytes(errpipe_data))
            child_exception_type = getattr(
                    builtins, exception_name.decode('ascii'),
                    SubprocessError)
            if issubclass(child_exception_type, OSError) and hex_errno:
                errno_num = int(hex_errno, 16)
                child_exec_never_called = (err_msg == "noexec")
                if child_exec_never_called:
                    err_msg = ""
                    # The error must be from chdir(cwd).
                    err_filename = cwd
                else:
                    err_filename = orig_executable
                if errno_num != 0:
                    err_msg = os.strerror(errno_num)
>               raise child_exception_type(errno_num, err_msg, err_filename)
E               FileNotFoundError: [Errno 2] No such file or directory: '/root/ocs-upi-kvm/src/ocs-ci/data/odf-cli'

/usr/lib64/python3.9/subprocess.py:1837: FileNotFoundError

The test occationally fails on 4.18 
Relevant 4.1 runs: 
https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/795/42553/1957761/1957938/log
https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/795/42556/1959057/1959156/log

Error: 
`self = <mcg.test_custom_credentials_using_mcg_cli.TestCustomCredsUsingMcgCli object at 0x7f84f829dd90>
mcg_account_factory = <function mcg_account_factory_fixture..mcg_account_factory_implementation at 0x7f851d5465c0>


@skipif_ocs_version("<4.17")
@tier2
def test_account_update_with_custom_creds_using_cli(self, mcg_account_factory):
    """
    1. Create a new account using MCG CLI
    2. Retrive creds of the account
    3. Change creds of the with custom value suing CLI
    4. Verify creds are updated using CLI
    5. Validate custom credentials against against length and valid characters
    """
    if not Path(mcg_cli).exists():
        retrieve_cli_binary(cli_type="mcg")

    output = run_cmd(cmd=f"{mcg_cli} version")
    logger.info(output)
    account_name = get_random_str(5)
    original_acc_credentials = mcg_account_factory(name=account_name)
    original_secret_key = original_acc_credentials["access_key"]
    original_access_key = original_acc_credentials["access_key_id"]

    logger.info(f"Original access key: {original_access_key}")
    logger.info(f"Original secret key: {original_secret_key}")
    logger.info("Updating noobaa account credentials with custom values")
    new_access_key = get_random_str(20)
    new_secret_key = get_random_str(40)
    self.update_nb_account(account_name, new_access_key, new_secret_key)
    retrived_access_key, retrived_secret_key = get_s3_credentials_from_secret(
        f"noobaa-account-{account_name}"
    )
    logger.info(f"retrived access key: {retrived_access_key}")
    logger.info(f"retrived secret key: {retrived_secret_key}")
    assert (
        retrived_access_key == new_access_key
    ), "Mismatch in updated and new access key"
    assert (
        retrived_secret_key == new_secret_key
    ), "Mismatch in updated and new secret key"
    logger.info("Custom credentials updated successfully")
    logger.info("Updating credentials with incorrect length")
    bad_length_access_key = get_random_str(19)
    bad_length_secret_key = get_random_str(39)
    self.update_nb_account(
        account_name, bad_length_access_key, bad_length_secret_key
    )
    retrived_access_key, retrived_secret_key = get_s3_credentials_from_secret(
        f"noobaa-account-{account_name}"
    )
    logger.info(f"retrived access key after bad value: {retrived_access_key}")
    logger.info(f"retrived secret key after bad value: {retrived_secret_key}")
    assert (
        retrived_access_key != bad_length_access_key
    ), "CLI is accepting access key with invalid lenght"
    assert (
        retrived_secret_key != bad_length_secret_key
    ), "CLI is accepting secret key with invalid lenght"
    logger.info("Credentials with incorrect length didnt updated as expected")
    logger.info("Updating credentials with invalid characters")
    bad_access_key = bad_length_access_key + "".join(
        random.choices(string.punctuation, k=1)
    )
    bad_secret_key = bad_length_secret_key + "".join(
        random.choices(string.punctuation.replace("+", "").replace("/", ""), k=1)
    )

  self.update_nb_account(account_name, bad_access_key, bad_secret_key)

/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/tests/functional/object/mcg/test_custom_credentials_using_mcg_cli.py:103: 



/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/tests/functional/object/mcg/test_custom_credentials_using_mcg_cli.py:31: in update_nb_account
    output = run_cmd(
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/utility/utils.py:491: in run_cmd
    completed_process = exec_cmd(
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/utility/retry.py:79: in f_retry
    return f(args, *kwargs)
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/utility/utils.py:663: in exec_cmd
    cmd = shlex.split(cmd)
/usr/lib64/python3.11/shlex.py:315: in split
    return list(lex)
/usr/lib64/python3.11/shlex.py:300: in next
    token = self.get_token()
/usr/lib64/python3.11/shlex.py:109: in get_token
    raw = self.read_token()



self = <shlex.shlex object at 0x7f850cbfb0d0>


def read_token(self):
    quoted = False
    escapedstate = ' '
    while True:
        if self.punctuation_chars and self._pushback_chars:
            nextchar = self._pushback_chars.pop()
        else:
            nextchar = self.instream.read(1)
        if nextchar == '\n':
            self.lineno += 1
        if self.debug >= 3:
            print("shlex: in state %r I see character: %r" % (self.state,
                                                              nextchar))
        if self.state is None:
            self.token = ''        # past end of file
            break
        elif self.state == ' ':
            if not nextchar:
                self.state = None  # end of file
                break
            elif nextchar in self.whitespace:
                if self.debug >= 2:
                    print("shlex: I see whitespace in whitespace state")
                if self.token or (self.posix and quoted):
                    break   # emit current token
                else:
                    continue
            elif nextchar in self.commenters:
                self.instream.readline()
                self.lineno += 1
            elif self.posix and nextchar in self.escape:
                escapedstate = 'a'
                self.state = nextchar
            elif nextchar in self.wordchars:
                self.token = nextchar
                self.state = 'a'
            elif nextchar in self.punctuation_chars:
                self.token = nextchar
                self.state = 'c'
            elif nextchar in self.quotes:
                if not self.posix:
                    self.token = nextchar
                self.state = nextchar
            elif self.whitespace_split:
                self.token = nextchar
                self.state = 'a'
            else:
                self.token = nextchar
                if self.token or (self.posix and quoted):
                    break   # emit current token
                else:
                    continue
        elif self.state in self.quotes:
            quoted = True
            if not nextchar:      # end of file
                if self.debug >= 2:
                    print("shlex: I see EOF in quotes state")
                # XXX what error should be raised here?

              raise ValueError("No closing quotation")
E                   ValueError: No closing quotation



/usr/lib64/python3.11/shlex.py:191: ValueError`

--- a/tests/functional/object/mcg/test_custom_credentials_using_mcg_cli.py
+++ b/tests/functional/object/mcg/test_custom_credentials_using_mcg_cli.py
@@ -1,6 +1,8 @@
 import logging
+import shlex
 import string
 import random
+from pathlib import Path
 
 from ocs_ci.framework.testlib import (
     skipif_ocs_version,
@@ -13,48 +15,72 @@ from ocs_ci.framework import config
 from ocs_ci.helpers.helpers import retrieve_cli_binary
 from ocs_ci.utility.utils import run_cmd, get_random_str
 from ocs_ci.helpers.helpers import get_s3_credentials_from_secret
-from pathlib import Path
 from ocs_ci.utility import version as version_module
-from ocs_ci.utility.utils import get_noobaa_cli_config
+from ocs_ci.ocs import constants
 
 
 logger = logging.getLogger(__name__)
-# Use version-appropriate CLI path and command prefix
-mcg_cli, mcg_command_prefix = get_noobaa_cli_config()
 
 
-def build_mcg_command(command_args):
+def ensure_mcg_cli():
+    """
+    Ensure the NooBaa CLI binary is downloaded and return
+    (cli_path, command_prefix).
+
+    For >= 4.20: uses ODFCLIRetriever → (bin_dir/odf, "noobaa")
+    For <  4.20: uses retrieve_cli_binary("mcg") → (DATA_DIR/mcg-cli, "")
+
+    Returns:
+        tuple: (cli_path, command_prefix)
+    """
+    ocs_version = version_module.get_semantic_ocs_version_from_config()
+
+    if ocs_version >= version_module.VERSION_4_20:
+        from ocs_ci.helpers.odf_cli import ODFCLIRetriever
+
+        retriever = ODFCLIRetriever()
+        if not retriever.check_odf_cli_binary():
+            retriever.retrieve_odf_cli_binary()
+        return retriever.local_cli_path, "noobaa"
+    else:
+        cli_path = constants.NOOBAA_OPERATOR_LOCAL_CLI_PATH
+        if not Path(cli_path).exists():
+            retrieve_cli_binary(cli_type="mcg")
+        return cli_path, ""
+
+
+def build_mcg_command(cli_path, prefix, command_args):
     """
     Build NooBaa CLI command with appropriate prefix based on version.
 
     Args:
+        cli_path (str): Path to the CLI binary
+        prefix (str): Command prefix ("noobaa" for odf-cli, "" for mcg-cli)
         command_args (str): Command arguments to append after CLI and prefix
 
     Returns:
         str: Complete CLI command string
     """
-    if mcg_command_prefix:
-        # For odf-cli: odf-cli noobaa <command>
-        return f"{mcg_cli} {mcg_command_prefix} {command_args}"
+    if prefix:
+        return f"{cli_path} {prefix} {command_args}"
     else:
-        # For mcg-cli: mcg-cli <command>
-        return f"{mcg_cli} {command_args}"
+        return f"{cli_path} {command_args}"
 
 
 @red_squad
 @mcg
 class TestCustomCredsUsingMcgCli(MCGTest):
-    def update_nb_account(self, account_name, access_key, secret_key):
+    def update_nb_account(self, cli_path, prefix, account_name, access_key, secret_key):
         """
         Update noobaa account with custom credential values
         """
         namespace = config.ENV_DATA["cluster_namespace"]
-
-        # Build command based on CLI version
-        cmd = build_mcg_command(f"account credentials {account_name}")
-
-        cmd += f" --access-key={access_key} --secret-key={secret_key} -n {namespace}"
-
+        cmd = build_mcg_command(cli_path, prefix, f"account credentials {account_name}")
+        cmd += (
+            f" --access-key={shlex.quote(access_key)}"
+            f" --secret-key={shlex.quote(secret_key)}"
+            f" -n {namespace}"
+        )
         output = run_cmd(cmd=cmd, ignore_error=True)
         logger.info(output)
 
@@ -64,23 +90,16 @@ class TestCustomCredsUsingMcgCli(MCGTest):
         """
         1. Create a new account using MCG CLI
         2. Retrive creds of the account
-        3. Change creds of the with custom value suing CLI
+        3. Change creds of the with custom value using CLI
         4. Verify creds are updated using CLI
-        5. Validate custom credentials against against length and valid characters
+        5. Validate custom credentials against length and valid characters
         """
-        if not Path(mcg_cli).exists():
-            # Download appropriate CLI binary based on version
-            ocs_version = version_module.get_semantic_ocs_version_from_config()
-            if ocs_version >= version_module.VERSION_4_20:
-                retrieve_cli_binary(cli_type="odf")
-            else:
-                retrieve_cli_binary(cli_type="mcg")
+        cli_path, prefix = ensure_mcg_cli()
 
-        # Build version command based on CLI version
-        version_cmd = build_mcg_command("version")
-
+        version_cmd = build_mcg_command(cli_path, prefix, "version")
         output = run_cmd(cmd=version_cmd)
         logger.info(output)
+
         account_name = get_random_str(5)
         original_acc_credentials = mcg_account_factory(name=account_name)
         original_secret_key = original_acc_credentials["access_key"]
@@ -91,7 +110,9 @@ class TestCustomCredsUsingMcgCli(MCGTest):
         logger.info("Updating noobaa account credentials with custom values")
         new_access_key = get_random_str(20)
         new_secret_key = get_random_str(40)
-        self.update_nb_account(account_name, new_access_key, new_secret_key)
+        self.update_nb_account(
+            cli_path, prefix, account_name, new_access_key, new_secret_key
+        )
         retrived_access_key, retrived_secret_key = get_s3_credentials_from_secret(
             f"noobaa-account-{account_name}"
         )
@@ -104,11 +125,12 @@ class TestCustomCredsUsingMcgCli(MCGTest):
             retrived_secret_key == new_secret_key
         ), "Mismatch in updated and new secret key"
         logger.info("Custom credentials updated successfully")
+
         logger.info("Updating credentials with incorrect length")
         bad_length_access_key = get_random_str(19)
         bad_length_secret_key = get_random_str(39)
         self.update_nb_account(
-            account_name, bad_length_access_key, bad_length_secret_key
+            cli_path, prefix, account_name, bad_length_access_key, bad_length_secret_key
         )
         retrived_access_key, retrived_secret_key = get_s3_credentials_from_secret(
             f"noobaa-account-{account_name}"
@@ -122,6 +144,7 @@ class TestCustomCredsUsingMcgCli(MCGTest):
             retrived_secret_key != bad_length_secret_key
         ), "CLI is accepting secret key with invalid lenght"
         logger.info("Credentials with incorrect length didnt updated as expected")
+
         logger.info("Updating credentials with invalid characters")
         bad_access_key = bad_length_access_key + "".join(
             random.choices(
@@ -133,7 +156,9 @@ class TestCustomCredsUsingMcgCli(MCGTest):
                 [ch for ch in string.punctuation if ch not in ['"', "'", "+", "/"]], k=1
             )
         )
-        self.update_nb_account(account_name, bad_access_key, bad_secret_key)
+        self.update_nb_account(
+            cli_path, prefix, account_name, bad_access_key, bad_secret_key
+        )
         retrived_access_key, retrived_secret_key = get_s3_credentials_from_secret(
             f"noobaa-account-{account_name}"
         )

--- a/tests/functional/object/mcg/test_custom_credentials_using_mcg_cli.py
+++ b/tests/functional/object/mcg/test_custom_credentials_using_mcg_cli.py
@@ -1,4 +1,5 @@
 import logging
+import os
 import shlex
 import string
 import random
@@ -44,7 +45,7 @@ def ensure_mcg_cli():
         return retriever.local_cli_path, "noobaa"
     else:
         cli_path = constants.NOOBAA_OPERATOR_LOCAL_CLI_PATH
-        if not Path(cli_path).exists():
+        if not (Path(cli_path).is_file() and os.access(cli_path, os.X_OK)):
             retrieve_cli_binary(cli_type="mcg")
         return cli_path, ""
 

